README.md
# [Load m Up](https://gitlab.com/GCSBOSS/load-m-up)

A RESTful service specialized in receiving and managing web file uploads.

## Get Started

1. Install with: `npm i -g --no-optional load-m-up`.
2. Optionally point `LOADMUP_CONF` env var to a TOML [config file](#configuration).
3. Optionally store a strong admin token in `LOADMUP_TOKEN` env var.
4. Run in the terminal with: `load-m-up`.

## Reporting Bugs
If you have found any problems with this module, please:

1. [Open an issue](https://gitlab.com/GCSBOSS/load-m-up/issues/new).
2. Describe what happened and how.
3. Also in the issue text, reference the label `~bug`.

We will make sure to take a look when time allows us.

## Proposing Features
If you wish to get that awesome feature or have some advice for us, please:
1. [Open an issue](https://gitlab.com/GCSBOSS/load-m-up/issues/new).
2. Describe your ideas.
3. Also in the issue text, reference the label `~proposal`.

## Contributing
If you have spotted any enhancements to be made and is willing to get your hands
dirty about it, fork us and
[submit your merge request](https://gitlab.com/GCSBOSS/load-m-up/merge_requests/new)
so we can collaborate effectively.

## API

Endpoint | Body | Visibility | Summary
---------|------|---------|---
`POST /upload` | `multipart/form-data` | Public | Upload files to the server
`POST /confirmation` | `application/json` | Admin | Mark an upload as permanent
`GET /upload/:hash/:name` | none | Public | Retrieve a given upload

## Configuration

The following options are available on the config file:

Option | Type | Summary | Default
-------|------|---------|--------
`debug`| Boolean | When true, enables some minor development env tweaks. | false
`port`| Boolean | HTTP port for the server to listen | 80 or 443
`confirmTimeout` | Integer | Timeout in seconds for unconfirmed uploads to be removed. | 6
`dir` | Directory | Where to store permanent uploads. | `./uploads`
`formFileDir` | Directory | Where to store unconfirmed uploads. | `./uploads/tmp`
`sizeLimit` | Integer | A maximum byte size for uploads. | No limit
`whitelist` | Array of String | An array of mime-types/extensions that should be accepted on uploads. | Disabled
`blacklist` | Array of String | An array of mime-types/extensions that should be rejected on uploads. | Disabled
`log.file` | File | Where to stream all log events. | Unset
`ssl.key` | File | Where to look for SSL key. | Unset
`ssl.cert` | File | Where to look for SSL cert/chain. | Unset
`multi` | Boolean | Whether to allow multiple files in 1 request | false
`multiSizeLimit` | Integer | A maximum overall byte size in 1 request | No Limit
`multiFileLimit` | Integer | A maximum amount of files allowed in 1 request | No Limit


CHANGELOG.md
# Load m Up Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.1.1] - 2019-07-25
- First officially published version.

## [v0.1.0] - 2019-07-15
- First officially published version.

[v0.1.0]: https://gitlab.com/GCSBOSS/load-m-up/-/tags/v0.1.0
[v0.1.1]: https://gitlab.com/GCSBOSS/load-m-up/-/tags/v0.1.1


test/spec.js
//const wtf = require('wtfnode');
const fs = require('fs');
const assert = require('assert');
const Tempper = require('tempper');
const FormData = require('form-data');
const { post, get } = require('muhb');

const delay = s => new Promise(done => setTimeout(done, s * 1000));
const parseBody = res => new Promise(resolve => {
    let body = '';
    res.on('data', data => body += data.toString());
    res.on('end', () => resolve(body));
});

const init = require('../lib/main');

const LOCAL_HOST = 'http://localhost:80/';

describe('Load m Up', function(){

    let tmp, app;

    before(function(){
        process.chdir('./test');
        tmp = new Tempper();
        tmp.mkdir('uploads/tmp');
    });

    after(function(){
        tmp.clear();
    });

    beforeEach(async function(){
        app = init();
        await app.start();
    });

    afterEach(async function(){
        await app.stop();
        tmp.refresh();
    });

    describe('Startup', function(){

        it('Should boot just fine', async function(){
            let { status } = await get(LOCAL_HOST);
            assert.strictEqual(status, 404);
        });

    });

    describe('GET /upload/:hash/:name', function(){

        it('Should return a not found error if file is not in the server', async function(){
            let { status, body } = await get(LOCAL_HOST + 'upload/hash/upload.txt');
            assert.strictEqual(status, 404);
            assert(/file doesn\'t/.test(body));
        });

        it('Should return the file stored in the server', async function(){
            tmp.mkdir('uploads/hash');
            tmp.addFile('./res/upload.txt', './uploads/hash/upload.txt');
            let { status, body } = await get(LOCAL_HOST + 'upload/hash/upload.txt');
            assert.strictEqual(status, 200);
            assert.strictEqual(body.replace(/\s+/g, ''), 'foobar');
        });

    });

    describe('POST /confirmation', function(){

        it('Should return an unauthorized error when token doesn\'t match', async function(){
            process.env.LOADMUP_TOKEN = 'notnone';
            let { status, body } = await post(
                LOCAL_HOST + 'confirmation',
                { 'Content-Type': 'application/json' },
                '{"token":"none"}'
            );
            assert.strictEqual(status, 401);
            assert(/Invalid access token/.test(body));
            delete process.env.LOADMUP_TOKEN;
        });

        it('Should return an invalid input error when temp upload doesn\'t exist', async function(){
            let { status, body } = await post(
                LOCAL_HOST + 'confirmation',
                { 'Content-Type': 'application/json' },
                '{"token":"none"}'
            );
            assert.strictEqual(status, 400);
            assert(/not found/.test(body));
        });

        it('Should copy temp upload to persistent dir', async function(){
            tmp.mkdir('uploads/tmp/my-hash');
            tmp.addFile('./res/upload.txt', './uploads/tmp/my-hash/upload.txt');
            let { status, body } = await post(
                LOCAL_HOST + 'confirmation',
                { 'Content-Type': 'application/json' },
                '{"token":"none","hash":"my-hash"}'
            );
            tmp.assertExists('uploads/my-hash/upload.txt');
            assert.strictEqual(status, 201);
            assert.strictEqual(body, '/my-hash/upload.txt');
        });

    });

    describe('POST /upload', function(){

        it('Should return bad request error when body IS NOT form-data', async () => {
            let { status, body } = await post(
                LOCAL_HOST + 'upload',
                { 'Content-Type': 'application/json' },
                '{"token":"none","hash":"my-hash"}'
            );
            assert.strictEqual(status, 400);
            assert(/Unsupported/.test(body));
        });

        it('Should return invalid input error when no body is sent', async () => {
            let { status, body } = await post(LOCAL_HOST + 'upload');
            assert.strictEqual(status, 400);
            assert(/Missing file field/.test(body));
        });

        it('Should store the sent file temporarily', function(done){
            this.timeout(8000);
            tmp.addFile('./res/upload.txt', 'upload.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'));
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                assert.strictEqual(res.statusCode, 201);
                assert.strictEqual(body.length, 64);
                tmp.assertExists('uploads/tmp/' + body + '/upload.txt');
                await delay(7);
                tmp.assertMissing('uploads/tmp/' + body + '/upload.txt');
                done();
            });
        });

        it('Should deny uploading multiple files', function(done){
            tmp.addFile('./res/upload.txt', 'upload.txt');
            tmp.addFile('./res/upload2.txt', 'upload2.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'));
            form.append('file', fs.createReadStream('./upload2.txt'));
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                assert.strictEqual(res.statusCode, 400);
                assert(/file limit of 1/.test(body));
                done();
            });
        });

    });

    describe('Settings', function(){

        it('Should remove tmp upload after setup amount of time [confirmTimeout] ', function(done){
            app.setup({ confirmTimeout: 1 });

            this.timeout(3000);
            tmp.addFile('./res/upload.txt', 'upload.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'));
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                tmp.assertExists('uploads/tmp/' + body + '/upload.txt');
                await delay(2);
                tmp.assertMissing('uploads/tmp/' + body + '/upload.txt');
                done();
            });
        });

        it('Should return the file stored in the setup dir [dir]', async function(){
            app.setup({ dir: './ups/' });

            tmp.mkdir('ups/my-hash');
            tmp.addFile('./res/upload.txt', './ups/my-hash/upload.txt');
            let { status, body } = await get(LOCAL_HOST + 'upload/my-hash/upload.txt');
            assert.strictEqual(status, 200);
            assert.strictEqual(body.replace(/\s+/g, ''), 'foobar');
        });

        it('Should copy temp upload to setup dir [dir] [formFileDir]', async function(){
            app.setup({ dir: './ups/', formFileDir: './tups' });

            tmp.mkdir('ups');
            tmp.mkdir('tups/my-hash');
            tmp.addFile('./res/upload.txt', './tups/my-hash/upload.txt');
            let { status, body } = await post(
                LOCAL_HOST + 'confirmation',
                { 'Content-Type': 'application/json' },
                '{"token":"none","hash":"my-hash"}'
            );
            tmp.assertExists('ups/my-hash/upload.txt');
            assert.strictEqual(status, 201);
            assert.strictEqual(body, '/my-hash/upload.txt');
        });

        it('Should store the sent file temporarily on setup dir [formFileDir]', function(done){
            app.setup({ formFileDir: './tups/', confirmTimeout: 1 });

            tmp.mkdir('tups');
            this.timeout(3000);
            tmp.addFile('./res/upload.txt', 'upload.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'));
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                assert.strictEqual(res.statusCode, 201);
                tmp.assertExists('tups/' + body + '/upload.txt');
                await delay(2);
                tmp.assertMissing('tups/' + body + '/upload.txt');
                done();
            });
        });

        it('Should deny uploads that exceed the setup size limit [sizeLimit]', function(done){
            app.setup({ sizeLimit: 3 });

            tmp.addFile('./res/upload.txt', 'upload.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'));
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                assert.strictEqual(res.statusCode, 400);
                assert(/Exceeded max/.test(body));
                done();
            });
        });

        it('Should deny mime-types not present on the whitelist [whitelist]', function(done){
            app.setup({ whitelist: [ 'application/json' ] });

            tmp.addFile('./res/upload.txt', 'upload.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'), { contentType: 'text/plain' });
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                assert.strictEqual(res.statusCode, 400);
                assert(/not allowed/.test(body));
                done();
            });
        });

        it('Should deny mime-types present on the blacklist [blacklist]', function(done){
            app.setup({ blacklist: [ 'text/plain' ] });

            tmp.addFile('./res/upload.txt', 'upload.txt');

            let form = new FormData();
            form.append('file', fs.createReadStream('./upload.txt'), { contentType: 'text/plain' });
            form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                let body = await parseBody(res);
                assert.strictEqual(res.statusCode, 400);
                assert(/not allowed/.test(body));
                done();
            });
        });

        describe('Multiple Files [multi]', function(){

            it('Should temporarily store multiple files sent', function(done){
                app.setup({ multi: true });

                this.timeout(8000);
                tmp.addFile('./res/upload.txt', 'upload.txt');
                tmp.addFile('./res/upload2.txt', 'upload2.txt');

                let form = new FormData();
                form.append('file', fs.createReadStream('./upload.txt'));
                form.append('file', fs.createReadStream('./upload2.txt'));
                form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                    let body = await parseBody(res);
                    assert.strictEqual(res.statusCode, 201);
                    assert.strictEqual(body.length, 64);
                    tmp.assertExists('uploads/tmp/' + body + '/upload.txt');
                    tmp.assertExists('uploads/tmp/' + body + '/upload2.txt');
                    await delay(7);
                    tmp.assertMissing('uploads/tmp/' + body + '/upload.txt');
                    tmp.assertMissing('uploads/tmp/' + body + '/upload2.txt');
                    done();
                });
            });

            it('Should copy tmp uploads to same persistent dir', async function(){
                app.setup({ multi: true });

                tmp.mkdir('uploads/tmp/my-hash');
                tmp.addFile('./res/upload.txt', './uploads/tmp/my-hash/upload.txt');
                tmp.addFile('./res/upload2.txt', './uploads/tmp/my-hash/upload2.txt');
                let { status, body } = await post(
                    LOCAL_HOST + 'confirmation',
                    { 'Content-Type': 'application/json' },
                    '{"token":"none","hash":"my-hash"}'
                );
                tmp.assertExists('uploads/my-hash/upload.txt');
                tmp.assertExists('uploads/my-hash/upload2.txt');
                assert.strictEqual(status, 201);
                assert.strictEqual(body, '/my-hash/upload.txt\r\n/my-hash/upload2.txt');
            });

            it('Should deny multi uploads whose overall size exceed the setup limit [multiSizeLimit]', function(done){
                app.setup({ multi: true, multiSizeLimit: 10 });

                tmp.addFile('./res/upload.txt', 'upload.txt');
                tmp.addFile('./res/upload2.txt', 'upload2.txt');

                let form = new FormData();
                form.append('file', fs.createReadStream('./upload.txt'));
                form.append('file', fs.createReadStream('./upload2.txt'));
                form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                    let body = await parseBody(res);
                    assert.strictEqual(res.statusCode, 400);
                    assert(/Exceeded overall/.test(body));
                    done();
                });
            });

            it('Should deny uploading more files than the setup limit [multiFileLimit]', function(done){
                app.setup({ multi: true, multiFileLimit: 1 });

                tmp.addFile('./res/upload.txt', 'upload.txt');
                tmp.addFile('./res/upload2.txt', 'upload2.txt');

                let form = new FormData();
                form.append('file', fs.createReadStream('./upload.txt'));
                form.append('file', fs.createReadStream('./upload2.txt'));
                form.submit(LOCAL_HOST + 'upload', async (err, res) => {
                    let body = await parseBody(res);
                    assert.strictEqual(res.statusCode, 400);
                    assert(/file limit of 1/.test(body));
                    done();
                });
            });

        });

    });

});

//process.on('SIGINT', () => wtf.dump());


lib/upload.js

const { resolve } = require('path');
const fs = require('fs').promises;
const { createReadStream } = require('fs');
const { createHash } = require('crypto');
const { valid, authorized } = require('nodecaf').assertions;

const
    TOO_BIG_MSG = 'Exceeded max file size allowed of %d bytes',
    ALL_TOO_BIG_MSG = 'Exceeded overall size limit of %d bytes',
    BLACKLISTED_MSG = 'MIME-TYPE %s is not allowed',
    NO_DATA_MSG = 'Missing file field on form data',
    NOT_ADMIN_MSG = 'Invalid access token',
    UNKNOWN_UPLOAD_MSG = 'Unconfirmed upload not found',
    NO_FILE_MSG = 'Requested file doesn\'t exist',
    TOO_MANY_MSG = 'Exceeded file limit of %d';

async function readDir(path){
    try{
        return await fs.readdir(path);
    }
    catch(e){
        return false;
    }
}

module.exports = {

    validate({ flash, conf, next }){
        flash.fileNames = '';

        for(let f of flash.files){

            let limit = conf.sizeLimit;
            valid(!limit || f.size <= limit, TOO_BIG_MSG, limit);

            let wl = conf.whitelist;
            valid(!wl || wl.includes(f.mimetype), BLACKLISTED_MSG, f.mimetype);

            let bl = conf.blacklist;
            valid(!bl || !bl.includes(f.mimetype), BLACKLISTED_MSG, f.mimetype);

            flash.fileNames += f.name;
        }
        next();
    },

    read({ flash, req, next, conf }){

        valid(req.files, NO_DATA_MSG);

        flash.files = [];
        for(let key in req.files)
            if(Array.isArray(req.files[key]))
                flash.files = flash.files.concat(req.files[key]);
            else
                flash.files.push(req.files[key]);

        let maxFiles = conf.multi ? conf.multiFileLimit : 1;
        valid(!maxFiles || flash.files.length <= maxFiles, TOO_MANY_MSG, maxFiles);

        if(conf.multi){
            let totalSize = flash.files.reduce( (a, c) => a + c.size, 0);
            let ok = !conf.multiSizeLimit || totalSize <= conf.multiSizeLimit;
            valid(ok, ALL_TOO_BIG_MSG, conf.multiSizeLimit);
        }

        next();
    },

    async store({ flash, res, conf }){

        let h = createHash('sha256').update(new Date() + flash.fileNames).digest('hex');
        let d = resolve(process.cwd(), conf.formFileDir, h);
        await fs.mkdir(d, { recursive: true });

        for(let f of flash.files)
            await fs.rename(resolve(process.cwd(), f.tempFilePath), d + '/' + f.name);

        setTimeout(async () => {
            for(let f of flash.files)
                await fs.unlink(d + '/' + f.name);
            await fs.rmdir(d);
        }, conf.confirmTimeout * 1000);

        res.status(201).end(h);
    },

    async confirm({ conf, body, res }){

        let token = process.env.LOADMUP_TOKEN || 'none';
        authorized(body.token == token, NOT_ADMIN_MSG);

        let d = resolve(process.cwd(), conf.formFileDir, body.hash || 'none');
        let files = await readDir(d);
        valid(files, UNKNOWN_UPLOAD_MSG);

        let p = resolve(process.cwd(), conf.dir, body.hash);
        await fs.mkdir(p, { recursive: true });

        let paths = [];
        for(let f of files){
            paths.push('/' + body.hash + '/' + f);
            await fs.copyFile(d + '/' + f, p + '/' + f);
        }

        res.status(201).end(paths.join('\r\n'));
    },

    load({ params, conf, res, error }){
        let f = resolve(process.cwd(), conf.dir, params.hash, params.name);

        let stream = createReadStream(f);
        stream.on('error', () => error('NotFound', NO_FILE_MSG));
        stream.on('open', () => stream.pipe(res));
    }

};
